### PR TITLE
E2E SSH Brute force attack

### PR DESCRIPTION
**Pull request by BelenValdivia**
|Related issue|
|---|
|Closes: #2889, #2932, #3004|

## Description

In this PR a test has been created to test the SSH Brute Force demo case. For this purpose we have created:

- New module: `end_to_end`.
- Common libraries: `event_monitor`.

Each test will include in its directory a `data` folder, where you will find the test configuration as playbooks, and the test itself.

## Configuration options

To launch these tests, we must indicate the path of the inventory:

```
python -m pytest -vvvs tests/end_to_end/test_brute_force/ --inventory_path=/home/belen/Repositories/wazuh-qa/tests/end_to_end/general_playbooks/inventory.yml 
```
It should be noted that, for local environments, it will be necessary to create the inventory for the moment. It should contain the following:

```
all:
  hosts:
    wazuh-manager:
      ansible_connection: ssh
      ansible_user: vagrant
      ansible_ssh_private_key_file: <path_to_private_key>
      ansible_python_interpreter: /usr/bin/python3
```

## Checks

- [x] Python codebase satisfies PEP-8 style style guide. `pycodestyle --max-line-length=120 --show-source --show-pep8 file.py`.
- [x] GitHub actions code checks

### CODE REVIEW

File: tests/end_to_end/test_brute_force/data/playbooks/credentials.yaml [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+            dest: /home/vagrant
+            remote_src: yes
+          become: True
+
+        - name: Get passwords file
+          fetch:
+            src: /home/vagrant/wazuh-install-files/passwords.wazuh
+            dest: /tmp/
+            flat: yes
+          become: True
```
Comment (jmv74211) [inline tests/end_to_end/test_brute_force/data/playbooks/credentials.yaml:18]:
```suggestion
- name: Get credentials
  hosts: wazuh-manager
  tasks:

    - name: Unzip wazuh install files
      unarchive:
        src: /home/vagrant/wazuh-install-files.tar
        dest: /home/vagrant
        remote_src: yes
      become: True

    - name: Get passwords file
      fetch:
        src: /home/vagrant/wazuh-install-files/passwords.wazuh
        dest: /tmp/
        flat: yes
      become: True
```

File: tests/end_to_end/test_brute_force/data/test_cases/cases_brute_force.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+- name: ssh_brute_force
+  description: Execute brute force attacks using SSH and check generated alerts
+  configuration_parameters: null
+  metadata:
+    rule:
+      id: 5712
+      level: 10
+      description: "sshd: brute force trying to get access to the system. Non existent user."
+      mitre_technique: "Brute Force"
```
Comment (juliamagan) [inline tests/end_to_end/test_brute_force/data/test_cases/cases_brute_force.yaml:10]:
```suggestion
  metadata:
    rule:
      id: 5712
      level: 10
      description: "sshd: brute force trying to get access to the system. Non existent user."
    extra:
      mitre_technique: "Brute Force"
```

File: tests/end_to_end/test_brute_force/test_brute_force.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
+'''
+copyright: Copyright (C) 2015-2022, Wazuh Inc.
+           Created by Wazuh, Inc. <info@wazuh.com>.
+           This program is free software; you can redistribute it and/or modify it under the terms of GPLv2
+type: e2e
+brief: Wazuh includes a registration process that provides the user with an automated mechanism to enroll agents with
+       minimal configuration steps. To register an agent using the enrollment method, a manager with a valid IP needs
+       to be configured first. The agent then checks for the registration key in the client.keys file, and when the file
+       is empty, it automatically requests the key from the configured manager the agent is reporting to.
```
Comment (juliamagan) [inline tests/end_to_end/test_brute_force/test_brute_force.py:9]: This brief doesn't describe this test or module.

File: tests/end_to_end/test_brute_force/data/playbooks/credentials.yaml (hunk 1 above)
Comment (jmv74211) [reply tests/end_to_end/test_brute_force/data/playbooks/credentials.yaml:18]: Don't tabulate the beginning. Check out the rest of the playbooks. Apply this to all files.

File: tests/end_to_end/test_brute_force/test_brute_force.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+'''
+
```
Comment (jmv74211) [inline tests/end_to_end/test_brute_force/test_brute_force.py:12]:
```suggestion
```

**Review by juliamagan — CHANGES_REQUESTED** [review]
GJ, but some changes are required.

**Review by jmv74211 — CHANGES_REQUESTED** [review]

File: tests/end_to_end/test_brute_force/test_brute_force.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import os
+import pytest
+from tempfile import gettempdir
+
+from wazuh_testing.tools import configuration as config
+from wazuh_testing.end_to_end import get_alert_indexer_api, make_query
+from wazuh_testing.event_monitor import check_event
```
Comment (juliamagan) [inline tests/end_to_end/test_brute_force/test_brute_force.py:19]:
```suggestion
from wazuh_testing.end_to_end import as e2e
from wazuh_testing.event_monitor import as evm
```

**Review by juliamagan — CHANGES_REQUESTED** [review]

File: tests/end_to_end/test_brute_force/data/test_cases/cases_brute_force.yaml (hunk 2 above)
Comment (BelenValdivia) [reply tests/end_to_end/test_brute_force/data/test_cases/cases_brute_force.yaml]: Done

File: tests/end_to_end/test_brute_force/test_brute_force.py (hunk 3 above)
Comment (BelenValdivia) [reply tests/end_to_end/test_brute_force/test_brute_force.py]: Done

File: tests/end_to_end/test_brute_force/data/playbooks/credentials.yaml (hunk 1 above)
Comment (BelenValdivia) [reply tests/end_to_end/test_brute_force/data/playbooks/credentials.yaml]: Done

File: tests/end_to_end/test_brute_force/test_brute_force.py (hunk 5 above)
Comment (BelenValdivia) [reply tests/end_to_end/test_brute_force/test_brute_force.py]: Done

File: tests/end_to_end/test_brute_force/test_brute_force.py (hunk 4 above)
Comment (BelenValdivia) [reply tests/end_to_end/test_brute_force/test_brute_force.py]: Done

**Comment by BelenValdivia** [comment]
## 2022/06/10

## Tests Results
| Test Path  | Results |
|--|--|
|  test/end_to_end/test_brute_force| 🟢 [R1-test-brute-force.zip](https://github.com/wazuh/wazuh-qa/files/8882039/R1-test-brute-force.zip) |
|  test/end_to_end/test_brute_force| 🟢 [R2-test-brute-force.zip](https://github.com/wazuh/wazuh-qa/files/8882040/R2-test-brute-force.zip) |
|  test/end_to_end/test_brute_force| 🟢 [R3-test-brute-force.zip](https://github.com/wazuh/wazuh-qa/files/8882041/R3-test-brute-force.zip)|

File: tests/end_to_end/test_brute_force/data/playbooks/credentials.yaml [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+        dest: /home/vagrant
+        remote_src: yes
+      become: True
+
+    - name: Get passwords file
+      fetch:
+        src: /home/vagrant/wazuh-install-files/passwords.wazuh
+        dest: /tmp/
+        flat: yes
+      become: True
```
Comment (juliamagan) [inline tests/end_to_end/test_brute_force/data/playbooks/credentials.yaml:17]: This playbook is no longer required

File: tests/end_to_end/test_brute_force/data/test_cases/cases_brute_force.yaml [hunk 7]
```diff
@@ -0,0 +1,10 @@
+- name: ssh_brute_force
+  description: Execute brute force attacks using SSH and check generated alerts
+  configuration_parameters: null
+  metadata:
+    rule:
+      id: 5712
+      level: 10
+      description: "sshd: brute force trying to get access to the system. Non existent user."
```
Comment (juliamagan) [inline tests/end_to_end/test_brute_force/data/test_cases/cases_brute_force.yaml:8]:
```suggestion
    rule.id: 5712
    rule.level: 10
    rule.description: "sshd: brute force trying to get access to the system. Non existent user."
```

File: tests/end_to_end/test_brute_force/test_brute_force.py [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+from wazuh_testing import end_to_end as e2e
+from wazuh_testing import event_monitor as evm
+
+# Test cases data
+alerts_json = os.path.join(gettempdir(), 'alerts.json')
+test_data_path = os.path.join(os.path.dirname(os.path.realpath(__file__)), 'data')
+test_cases_file_path = os.path.join(test_data_path, 'test_cases', 'cases_brute_force.yaml')
+
+# Playbooks
+configuration_playbooks = ['credentials.yaml']
```
Comment (juliamagan) [inline tests/end_to_end/test_brute_force/test_brute_force.py:15]:
```suggestion

```

File: tests/end_to_end/test_brute_force/test_brute_force.py [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+    Test to detect a SSH Brute Force attack
+    """
+    rule_id = metadata['rule']['id']
+    rule_level = metadata['rule']['level']
+    rule_description = metadata['rule']['description']
+    rule_mitre_technique = metadata['extra']['mitre_technique']
+
+    expected_alert = r'\{{"timestamp":"(\d+\-\d+\-\w+\:\d+\:\d+\.\d+\+\d+)","rule"\:{{"level"\:{},"description"\:"{}",'\
+                     r'"id"\:"{}".*\}}'.format(rule_level, rule_description, rule_id)
+
```
Comment (juliamagan) [inline tests/end_to_end/test_brute_force/test_brute_force.py:35]: We need another regex that matches the indexed alert, take a look at:

https://github.com/wazuh/wazuh-qa/blob/81d6c7259522275c88767e0fb531661e9a3036a4/tests/end_to_end/test_audit/test_audit.py#L35-L39

File: tests/end_to_end/test_brute_force/test_brute_force.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+        assert indexed_alert['hits']['hits'][0]['_source']['rule']['id'] == str(rule_id), 'Invalid rule id'
+        # Check that indexed alert has the correct rule level
+        assert indexed_alert['hits']['hits'][0]['_source']['rule']['level'] == rule_level, 'Invalid rule level'
+        # Check that indexed alert has the correct rule description
+        assert indexed_alert['hits']['hits'][0]['_source']['rule']['description'] == rule_description, 'Invalid description'
+        # Check that indexed alert has the correct mitre technique
+        assert indexed_alert['hits']['hits'][0]['_source']['rule']['mitre']['technique'][0] == rule_mitre_technique, 'Invalid mitre technique'
+    except AssertionError:
+        evm.check_event(callback=expected_alert, file_to_monitor=alerts_json, error_message='The alert has not occurred')
+        raise AssertionError('The alert has occurred, but has not been indexed.')
```
Comment (juliamagan) [inline tests/end_to_end/test_brute_force/test_brute_force.py:56]: We need to follow a standard, take a look at: https://github.com/wazuh/wazuh-qa/blob/81d6c7259522275c88767e0fb531661e9a3036a4/tests/end_to_end/test_audit/test_audit.py#L55-L77

**Review by juliamagan — CHANGES_REQUESTED** [review]
GJ, but some changes are required

File: tests/end_to_end/test_brute_force/data/test_cases/cases_brute_force.yaml (hunk 7 above)
Comment (BelenValdivia) [reply tests/end_to_end/test_brute_force/data/test_cases/cases_brute_force.yaml]: Done in 85a36ccf7e801c47cce59c4f3fa7a90ec0ac563f

File: tests/end_to_end/test_brute_force/test_brute_force.py (hunk 8 above)
Comment (BelenValdivia) [reply tests/end_to_end/test_brute_force/test_brute_force.py]: Done in 85a36ccf7e801c47cce59c4f3fa7a90ec0ac563f

**Comment by BelenValdivia** [comment]
## 2022/06/10

## Tests Results
| Test Path  | Results |
|--|--|
|  tests/end_to_end/test_brute_force| 🟢 [R1-test-brute-force.zip](https://github.com/wazuh/wazuh-qa/files/8911203/R1-test-brute-force.zip) |
|  tests/end_to_end/test_brute_force| 🟢 [R2-test-brute-force.zip](https://github.com/wazuh/wazuh-qa/files/8911202/R2-test-brute-force.zip) |
|  tests/end_to_end/test_brute_force| 🟢 [R3-test-brute-force.zip](https://github.com/wazuh/wazuh-qa/files/8911200/R3-test-brute-force.zip) |

**Review by juliamagan — APPROVED** [review]

**Review by jmv74211 — APPROVED** [review]

**Pull request merged**